rulesets/src/collections.ruleset.yml

```diff
@@ -0,0 +1,136 @@
+
+  ##### Pagination #####
+  sps-missing-pagination-query-parameters:
+    description: 'Collection GET endpoints must support pagination using query parameters: limit, offset, cursor.'
+    severity: error
+    given: $.paths[?(!@property.match(/.*\/\{[^}]+\}$/))].get
+    then:
+      - field: parameters
+        function: schema
+        functionOptions:
+          schema:
+            type: array
+            items:
+              type: object
+              properties:
+                name:
+                  type: string
+              required:
+                - name
+            minItems: 3
+            contains:
+              anyOf:
+                - properties:
+                    name:
+                      const: offset
+                - properties:
+                    name:
+                      const: limit
+                - properties:
+                    name:
+                      const: cursor
+
+  sps-post-request-body-missing-paging-object:
+    description: "POST collection endpoints MUST have a request body schema that includes paging parameters."
+    severity: error
+    given: $.paths[?(!@property.match(/.*\/\{[^}]+\}$/))].post.requestBody.content.application/json.schema.properties.paging
+    then:
+      field: "type"
+      function: pattern
+      functionOptions:
+        match: "object"
+
+  ##### FILTERING #####
+  sps-disallow-resource-identifier-filtering:
+    description: "Resource identifier filtering is not allowed. Use the resource identifier in the URL path."
+    severity: warn
+    given: $.paths..get.parameters.[?(@.in=='query' && @.name=='id')]
+    then:
+      field: "name"
+      function: pattern
+      functionOptions:
+        notMatch: "^id$"
+
+  sps-unreasonable-query-parameters-limit:
+    description: "Filtering query parameters SHOULD have a reasonable limit, no more than 12."
+    severity: warn
+    given: $.paths..get
+    then:
+      function: schema
+      functionOptions:
+        dialect: "draft2020-12"
+        schema:
+          type: object
+          properties:
+            parameters:
+              type: array
+              minContains: 0
+              maxContains: 12
+              contains:
+                type: object
+                properties:
+                  in:
+                    const: query
+
+  sps-multiple-filter-parameters:
+    description: "Ensures the 'filter' query parameter is specified only once."
+    severity: error
+    given: $.paths..get
+    then:
+      function: schema
+      functionOptions:
+        dialect: "draft2020-12"
+        schema:
+          type: object
+          properties:
+            parameters:
+              type: array
+              minContains: 0
+              maxContains: 1
+              contains:
+                type: object
+                properties:
+                  in:
+                    const: query
+                  name:
+                    pattern: "(filter|Filter)"
```

Comment:
is this valid in openapi spec to specify the same query parameter with the same name twice? If not, then we don't need a rule for it